test/e2e/dnslookup_test.go

```diff
@@ -0,0 +1,299 @@
+}
+
+// TODO (Nino-K): maybe this can be replaced by CI
+func buildBinaries(path, goos, tmpDir string) error {
+	buildCmd := exec.Command("go", "build", "-o", tmpDir, path)
+	buildCmd.Env = os.Environ()
+	buildCmd.Env = append(buildCmd.Env, fmt.Sprintf("GOOS=%s", goos))
+	buildCmd.Stdout = os.Stdout
+	buildCmd.Stderr = os.Stderr
+
+	return buildCmd.Run()
+}
+
+func cmdRunWithOutput(command string, args ...string) (string, error) {
+	var outBuf, errBuf strings.Builder
+	cmd := exec.Command(command, args...)
+	cmd.Stdout = &outBuf
+	cmd.Stderr = &errBuf
+	err := cmd.Run()
+	if err != nil {
+		return errBuf.String(), err
+	}
+	return outBuf.String(), nil
+}
+
+func cmdExec(execDir, command string, args ...string) *exec.Cmd {
+	cmd := exec.Command(command, args...)
+	if execDir != "" {
+		cmd.Dir = execDir
+	}
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	return cmd
+}
+
+func copyTestData(src, dst string) error {
+	bytesRead, err := os.ReadFile(src)
+	if err != nil {
+		return err
+	}
+	return os.WriteFile(dst, bytesRead, 0600)
+}
+
+func cacheExsitingDNSAddrs(adapterAddrs []*winipcfg.IPAdapterAddresses) (map[string][]netip.Addr, error) {
+	guidToDNSAddrs := make(map[string][]netip.Addr)
+	for _, a := range adapterAddrs {
+		guid, err := a.LUID.GUID()
+		if err != nil {
+			return nil, err
+		}
+		dnsAddrs, err := a.LUID.DNS()
+		if err != nil {
+			return nil, err
+		}
+		guidToDNSAddrs[guid.String()] = dnsAddrs
+	}
+	return guidToDNSAddrs, nil
+}
+
+func confirm(command func() bool, interval, timeout time.Duration) error {
+	tick := time.NewTicker(interval)
+	terminate := time.After(timeout)
+
+	for {
+		select {
+		case <-tick.C:
+			if command() {
+				return nil
+			}
+		case <-terminate:
+			return fmt.Errorf("Failed to run within %v", timeout)
+		}
+	}
+}
+
+func updateSystemDNS(t *testing.T, dnsInfs []*winipcfg.IPAdapterAddresses) {
+	testDNSAddr := netip.MustParseAddr(testSrvAddr)
+	testDNSAddrIPv6 := netip.IPv6Unspecified()
+	for _, addr := range dnsInfs {
+		// Set IPv4 DNS
+		err := addr.LUID.SetDNS(windows.AF_INET, []netip.Addr{testDNSAddr}, []string{})
+		require.NoErrorf(t, err, "Failed setting IPv4 DNS server for: %v", addr.FriendlyName())
+		// Set IPv6 DNS to unspecified so DNS lookup will not be bypassed
+		err = addr.LUID.SetDNS(windows.AF_INET6, []netip.Addr{testDNSAddrIPv6}, []string{})
+		require.NoErrorf(t, err, "Failed setting IPv6 DNS server for: %v", addr.FriendlyName())
+	}
+}
+
+func restoreSystemDNS(t *testing.T, addrs []*winipcfg.IPAdapterAddresses, cachedAddrs map[string][]netip.Addr) {
+	t.Log("Restoring DNS servers back to the original state")
+	for _, addr := range addrs {
+		err := addr.LUID.FlushDNS(windows.AF_INET)
+		require.NoErrorf(t, err, "Failed to flush DNS for IPv4 addrs for: %v", addr.FriendlyName())
+		err = addr.LUID.FlushDNS(windows.AF_INET6)
+		require.NoErrorf(t, err, "Failed to flush DNS for IPv6 addrs for: %v", addr.FriendlyName())
+	}
+}
+
+func downloadWSLTarball(path, url string) error {
```

Comment:
Nothing in this function actually cares that this is a tarball.
```suggestion
func downloadFile(path, url string) error {
```